fem/normal_deriv_restriction.cpp

```diff
@@ -0,0 +1,854 @@
+            case 7: Mult2D<7>(x, y); break;
+            case 8: Mult2D<8>(x, y); break;
+            default: Mult2D(x, y); break;
+         }
+      }
+      break;
+      case 3:
+      {
+         const int d1d = fes.GetElementOrder(0) + 1;
+         switch (d1d)
+         {
+            case 1: Mult3D<1>(x, y); break;
+            case 2: Mult3D<2>(x, y); break;
+            case 3: Mult3D<3>(x, y); break;
+            case 4: Mult3D<4>(x, y); break;
+            case 5: Mult3D<5>(x, y); break;
+            case 6: Mult3D<6>(x, y); break;
+            case 7: Mult3D<7>(x, y); break;
+            case 8: Mult3D<8>(x, y); break;
+            default: Mult3D(x, y); break; // fallback
+         }
+         break;
+      }
+      default: MFEM_ABORT("Dimension not supported."); break;
+   }
+}
+
+void L2NormalDerivativeFaceRestriction::AddMultTranspose(
+   const Vector &x, Vector &y, const real_t a) const
+{
+   if (nf == 0) { return; }
+   switch (dim)
+   {
+      case 2:
+      {
+         const int d1d = fes.GetElementOrder(0) + 1;
+         switch (d1d)
+         {
+            case 1: AddMultTranspose2D<1>(x, y, a); break;
+            case 2: AddMultTranspose2D<2>(x, y, a); break;
+            case 3: AddMultTranspose2D<3>(x, y, a); break;
+            case 4: AddMultTranspose2D<4>(x, y, a); break;
+            case 5: AddMultTranspose2D<5>(x, y, a); break;
+            case 6: AddMultTranspose2D<6>(x, y, a); break;
+            case 7: AddMultTranspose2D<7>(x, y, a); break;
+            case 8: AddMultTranspose2D<8>(x, y, a); break;
+            default: AddMultTranspose2D(x, y, a); break;
+         }
+      }
+      break;
+      case 3:
+      {
+         const int d1d = fes.GetElementOrder(0) + 1;
+         switch (d1d)
+         {
+            case 1: AddMultTranspose3D<1>(x, y, a); break;
+            case 2: AddMultTranspose3D<2>(x, y, a); break;
+            case 3: AddMultTranspose3D<3>(x, y, a); break;
+            case 4: AddMultTranspose3D<4>(x, y, a); break;
+            case 5: AddMultTranspose3D<5>(x, y, a); break;
+            case 6: AddMultTranspose3D<6>(x, y, a); break;
+            case 7: AddMultTranspose3D<7>(x, y, a); break;
+            case 8: AddMultTranspose3D<8>(x, y, a); break;
+            default: AddMultTranspose3D(x, y, a); break; // fallback
+         }
+         break;
+      }
+      default: MFEM_ABORT("Not yet implemented"); break;
+   }
+}
+
+template <int T_D1D>
+void L2NormalDerivativeFaceRestriction::Mult2D(const Vector &x, Vector &y) const
+{
+   int ne_shared = 0;
+   const real_t *face_nbr_data = nullptr;
+
+#ifdef MFEM_USE_MPI
+   std::unique_ptr<ParGridFunction> x_gf;
+   if (const auto *pfes = dynamic_cast<const ParFiniteElementSpace*>(&fes))
+   {
+      if (face_type == FaceType::Interior)
+      {
+         x_gf.reset(new ParGridFunction(const_cast<ParFiniteElementSpace*>(pfes),
+                                        const_cast<Vector&>(x), 0));
+         x_gf->ExchangeFaceNbrData();
```

Comment:
It looks like we do this same exchange two times for DG diffusion. Can we factor it out to avoid the repetition?

Reply:
You mean that it's done once in `ParL2FaceRestriction` and once in `L2NormalDerivativeFaceRestriction`?

It would be good to get rid of this repetition. Do you have an idea how to structure it?

Reply:
Yes, the other place is here: https://github.com/mfem/mfem/blob/88a79f77889cdf59a306959157a3ea40b2e6360d/fem/prestriction.cpp#L317

It seems natural to do the exchange in `PABilinearFormExtension::Mult` and pass the `ParGridFunction` as input to two new methods that use it to compute `int_face_X` and `int_face_dXdn`. I'm not sure if this will have noticeable impact on performance, so it may not be worth it. What do you think?

Reply:
OK, I implemented something in 936d12883 (and minor change in 54c186eec).

There is a new helper function called `GetLVectorFaceNbrData`. This function checks if the input `Vector&` is actually a ParGridFunction. If so, and if the face-neighbor data is not empty, it just returns an alias to the existing face-neighbor data, without exchanging.

Otherwise, it does the exchange and returns the vector.

This way, if ParGridFunctions are passed into to `ParL2FaceRestriction` or `L2NormalDerivativeFaceRestriction`, the face-neighbor data can be retrieved without communication.

`PABilinearFormExtension::Mult` creates a temporary ParGridFunction (aliasing the input L-vector `x`), which is used to cache the face-neighbor data.